opacclient/opacapp/src/main/java/de/geeksfactory/opacclient/frontend/StarredFragment.java

```diff
@@ -256,6 +290,166 @@ protected void export() {
+        try {
+            starred.put(JSON_LIBRARY_NAME, app.getLibrary().getIdent());
+            JSONArray items = new JSONArray();
+            StarDataSource data = new StarDataSource(getActivity());
+            List<Starred> libItems = data.getAllItems(app.getLibrary().getIdent());
+            for (Starred libItem : libItems) {
+                JSONObject item = new JSONObject();
+                item.put(JSON_ITEM_MNR, libItem.getMNr());
+                item.put(JSON_ITEM_TITLE, libItem.getTitle());
+                item.put(JSON_ITEM_MEDIATYPE, libItem.getMediaType());
+                items.put(item);
+            }
+            starred.put(JSON_STARRED_LIST, items);
+            //write out
+
+            PrintWriter pw = new PrintWriter(dest, true);
+            pw.write(starred.toString());
+            pw.close();
+
+        } catch (JSONException e) {
+            showExportError();
+        }
+    }
+
+    public void importFromStorage() {
+        //Use SAF
+        Intent intent = null;
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT) {
+            intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
+            intent.addCategory(Intent.CATEGORY_OPENABLE);
+            intent.setType("application/json");
+        } else {    //let user use a custom picker
+            intent = new Intent(Intent.ACTION_GET_CONTENT);
+            intent.addCategory(Intent.CATEGORY_OPENABLE);
+            intent.setType("application/json");
+        }
+        try{
+            startActivityForResult(intent, REQUEST_CODE_IMPORT);
+        }catch (ActivityNotFoundException e) {
+            showImportError();//No picker app installed!
+        }
+    }
+
+    @Override
+    public void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (requestCode == REQUEST_CODE_EXPORT && resultCode == Activity.RESULT_OK) {
+            Log.i("StarredFragment", data.toString());
+            Uri uri = data.getData();
+            try {
+                OutputStream os = getActivity().getContentResolver().openOutputStream(uri);
+                if (os != null) {
+                    writeToFile(os);
+                    os.close();
+                } else {
+                    showExportError();
+                }
+            } catch (FileNotFoundException e) {
+                showExportError();
+            } catch (IOException e) {
+                showExportError();
+            }
+        } else if (requestCode == REQUEST_CODE_IMPORT && resultCode == Activity.RESULT_OK) {
+            Uri uri = data.getData();
+            try {
+                StarDataSource dataSource = new StarDataSource(getActivity());
+                InputStream is = getActivity().getContentResolver().openInputStream(uri);
+                if(is!=null) {
+                    BufferedReader reader = new BufferedReader(new InputStreamReader(is, "UTF-8"));
+                    String line = reader.readLine();
```

Comment:
My assumption was that the app wouldn't come across a JSON file that does not conform to its (private) format (since the app itself creates the JSON in the first place).